bus_analyzer/bus_analyzer/extract.py

```diff
@@ -0,0 +1,201 @@
+
+import os
+from io import BytesIO
+
+import argh
+from PIL import Image, ImageStat
+
+from common.segments import extract_frame, parse_segment_path
+
+
+# DB2023 buscam
+# bounding box (left x, top y, right x, bottom y) of the area the odometer can be in
+ODO_COORDS = 1121, 820, 1270, 897
+# starting x coord of each digit within the odo box
+DIGIT_X_COORDS = [0, 28, 56, 84, 123]
+DIGIT_WIDTH = 26
+# Most digits we only care about the actual character height
+DIGIT_HEIGHT = 26
+# But last digit we want the full white-background area as we want to try to match
+# based on position also.
+LAST_DIGIT_HEIGHT = 38
+
+# get back py2 zip behaviour
+_zip = zip
+def zip(*args):
+	return list(_zip(*args))
+
+
+cli = argh.EntryPoint()
+
+@cli
+@argh.arg("paths", nargs="+")
+def to_digits(output_dir, paths, box_only=False):
+	"""Extracts each digit and saves to a file. Useful for testing or building prototypes."""
+	if not os.path.exists(output_dir):
+		os.mkdir(output_dir)
+	for path in paths:
+		name = os.path.splitext(os.path.basename(path))[0]
+		image = Image.open(path)
+		if not box_only:
+			image = extract_odo(image)
+		for i, digit in enumerate(extract_digits(image)):
+			output_path = os.path.join(output_dir, "{}-digit{}.png".format(name, i))
+			digit.save(output_path)
+
+
+def extract_odo(image):
+	"""Takes a full frame, and returns the odo box"""
+	return image.crop(ODO_COORDS)
+
+
+def extract_digits(image, include_last=True):
+	"""Takes an odo box, and returns a list of 9x6 digit images"""
+	# convert to greyscale
+	image = image.convert(mode='L')
+	digits = []
+	for i, x in enumerate(DIGIT_X_COORDS):
+		# last digit is special
+		is_last = i == len(DIGIT_X_COORDS) - 1
+		if is_last and not include_last:
+			continue
+		digit = image.crop((x, 0, x + DIGIT_WIDTH, image.height))
+		digits.append(normalize_digit(digit, is_last))
+	return digits
+
+
+def normalize_digit(digit, is_last=False):
+	# Calculate total brightness by row
+	rows = [
+		sum(digit.getpixel((x, y)) for x in range(digit.width))
+		for y in range(digit.height)
+	]
+	# Find brightest sub-image of DIGIT_HEIGHT rows
```

Comment:
Does the odometer only move by integer numbers of pixels?

Reply:
Given we're dealing with an upscaled version of a low-resolution image, to a first approximation we can treat it like it does. The edges are already a little fuzzy.